### ISSUE TITLE
Improve external keyboard support

### ISSUE THREAD

**Issue by bwhelm** [terrychou/iVim#3]
Vim, of course, works best with a hardware keyboard. Mostly it works well, but special keys like `<Ctrl>` and `<Esc>` are not recognized properly. 

- `<Esc>` doesn't seem to work at all: I must use the software keyboard to enter this.
- `<Ctrl>` seems to register (so that `<C-a>` is different from `a`), but `<C-a>` seems to do nothing. For example, if in normal mode I type `<C-a>` over a number, nothing happens, though typing just `a` does what's expected.

I should add that I normally use the Dvorak layout on my external keyboard, but the problem remains if I switch back to the standard US keyboard. (Incidentally, one issue with other iOS apps with Dvorak keyboards is that entering <Ctrl> codes ignores the Dvorak layout. For example, hitting the `s` key when in Dvorak layout will enter `o`, but holding down the `Ctrl` key while hitting the `s` key wrongly enters `<C-s>` when it should enter `<C-o>`.)

**Comment by terrychou** [comment]
For now, there is no direct support (no code to handle) for external keyboards input in iVim because it might conflict with the multistage inputs. It may be added in coming versions.

**Comment by mvolkmann** [comment]
Adding support the for the esc key on external keyboards would be huge feature!

**Comment by terrychou** [comment]
I am implementing the external keyboard support now, would you people like to test it with TestFlight? Please offer me your emails.

**Comment by mvolkmann** [comment]
Sure, I'll test it! r.mark.volkmann@gmail.com

**Comment by terrychou** [comment]
The beta version has been sent to your emails, if I did not misspell them :).

**Comment by bwhelm** [comment]
I've tested the beta with iOS 11 and an Apple keyboard, using Dvorak key mapping. Here's what I found:

1. The most common keys seem to be working: `<Esc>`, standard arrow keys, `<C-[lowercase letter]>`, `<C-[>`, `<C-]>`, `<C-\>`, and all `<Opt-[key]>`.
2. What's not working: `<C-CR>`, `<C-Tab>`, and `<C-[arrow keys]>`.
3. `<Cmd-[key]>` doesn't seem to be recognized at all, though since that's not well supported by Vim, I don't know if it's worth trying to get it to work.

In short, aside from (2), everything seems to be working very nicely. Thanks!

Extra request: might it be possible to swap around keys like `<Esc>`, `<Caps Lock>`, and `<Ctrl>`? Blink.app (an open-source mosh client for iOS) does it: <https://github.com/blinksh/blink>. (I can submit this as a separate issue, if you'd like.)

**Comment by mvolkmann** [comment]
I just tested the esc key. Works great! Thanks so much for adding this!

Can this be used to edit files on Dropbox? How about files in a GitHub repo?

---
R. Mark Volkmann
Object Computing, Inc.

**Comment by bwhelm** [comment]
(On Dropbox and GitHub: see `:h idoc` for details. You'll of course need an iOS git client; I can recommend Working Copy.app, but there may well be others.)

**Comment by bwhelm** [comment]
Playing around with this a bit more, I notice that keys do not repeat. That would be nice to have on an external keyboard.

**Comment by terrychou** [comment]
@bwhelm What did you mean by "swap around keys"? Could you give an example?

As to all the expected keys and mappings you listed in the testing result, which platform are they based on? gvim? For example, <C-Tab> seems not working in vim in a shell.

**Comment by sedm0784** [comment]
I hesitated to post this before, because `has("gui_running")` returns true in iVim, so I presumed it was a GUI Vim rather than a terminal Vim, but `<C-Tab>` *cannot* work in terminal Vim, and `<C-CR>` and `<C-arrows>` don't work reliably, because terminals handle them in different ways.

[This article](https://catern.com/posts/terminal_quirks.html) is a good introduction to some of the issues.

**Comment by sedm0784** [comment]
> What did you mean by "swap around keys"? Could you give an example?

Many Vim users like to remap their Caps Lock key to perform some other function, such as Esc or Ctrl ([or both](https://gist.github.com/sedm0784/4443120)!)

**Comment by bwhelm** [comment]
Yes: that's what I meant. So you press the `<CapsLock>` key, and Vim interprets it as `<Esc>` or as `<Ctrl>`.

**Comment by terrychou** [comment]
Ok, you just meant to "remap the `caps lock` key", not swapping it with another one. It will be implemented.

Also, what other key combinations do you expect? like besides a single modifier key, do you expect multiple ones such as `<D-C-Tab>`? Is there a list of key combinations that vim supports?

**Comment by bwhelm** [comment]
To be honest, I don't understand all the interactions between Vim and various terminals and guis. Depending on what I need to do or what computer I'm on, I use vim (or neovim) in the terminal (with or without tmux), or with the gui (MacVim or VimR), and some of these allow for broader ranges of key combinations than others. As a result, I tend to go with what I know will work in all of them, which means I don't use key chords like `<D-C-Tab>` or `<C-S-[letter]>`.

Personally I map `<C-CR>`, `<S-Tab>`, and `<C-[key]>` where "[key]" includes "a-z]\/", but I'm sure others have other preferences given the terminal or gui environments in which they use Vim. I'm sure the more you're able to implement (including chords), the happier some people will be. On the other hand, I think `<Option>`/`<Alt>` mappings should probably be left alone since they have standard uses in iOS to produce things like accented characters and special symbols, but that's surely up for debate given that Vim has its own ways of producing those characters.

One other thing I've noticed: currently the "Delete" key (as it is labeled on Mac keyboards; it is also sometimes labeled as "Back" or "Backspace" and is distinct from the forward "Del" key) is interpreted as `<C-h>`. I'd prefer to have it interpreted as `<BS>`, which is consistent with other terminals I use and which frees up `<C-h>` to be mapped to something else.

**Comment by terrychou** [comment]
Just submitted the 2nd build of the beta version via TestFlight.

However, this build is supposed to be buggy:

The external keyboard related part had been almost rewritten.

1. now the chords include `D-key`, `A-key`,`C-key`, key means all alphabets, numerics, symbols, tab, enter, esc, arrows. It is weird `C-CR` never worked on my side.

2. there is no multiple modifiers chords

3. added the `caps lock` mapping part. But I would call it a failure. Because there are problems: a. the single `caps lock` can be triggered, which means the `esc` mapping would work. But the triggered `caps lock` key would make following alphabetical input all uppercased; b. as to the `ctrl` mapping, the `caps lock-key` seems not working on my side.

4. as to the key repeat. I had figured out a way but it would conflict with the multistage input, so just dropped it.

5. as to the delete key. I have the solution but it has been included in this build.

Please test them and tell me what it is like on your sides.

**Comment by bwhelm** [comment]
1. `D-key` and `C-key` seem to be working in most cases. Thus if I nmap `D-a` to `ihello<Esc>`, and then hit "<D-a>" in normal mode, "hello" is inserted, as expected. This seems to work for all single keymaps I've tried. However:
    1. `A-key` does not work as I'd expect. (See my comment above: I would leave `<Alt>` mappings alone because they have standard uses on iOS. But setting this aside....) In insert mode, when I type `A-e`, "å" shows up on the screen; if I then type a single `e`, that comes out as "é". On iOS, what I would expect is that typing `A-e` would not enter a character on its own but would modify the next character I type by adding an acute accent, so that `A-e` followed by `e` should produce "é". I would also expect to type "å" by entering `A-a`. So what I'm seeing on typing `A-e` is some weird combination of `A-a` and `A-e`. This is true of all `A-key` keys I've tried.
    2. I noted above (and now see issue #17 also raises this) that `Delete` is interpreted as `C-h` rather than as `BS`. Notice, however, that `CR` is interpreted as `C-m`, which is why you're finding that `C-CR` isn't working: it doesn't make sense to hit `Ctrl` twice. So if you've found a solution to the `Delete` problem, I'm guessing something analogous will work for `CR` as well.
    3. Notice that I said `C-key` and `D-key` work for *single* keymaps. I have `<C-t>m` mapped to one function and `<C-t>t` mapped to another. However, these do not work for me: simply pressing `C-t` is interpreted by iVim as the end of the command, and so I get the "E73: Tag stack empty" message, rather than having Vim wait for a second key. (Weirdly, if I do `:messages`, where Vim prints out something long at the bottom of the screen followed by "Press ENTER or type command to continue", and I then do `<C-t>m`, it works as expected.)

2. Right.

3. `Caps-lock` is not working well, as you report. I see the same behavior you report when I map `Caps-lock` to `esc`. When I map `Caps-lock` to `ctrl`, however, it looks like iVim is initially interpreting a press of `Caps-lock` as "<C->", and then waiting for a second key press. For example, I have `inoremap <C-b> <C-o>b`. In insert mode, when I press (and hold) `Caps-lock`, "<C->" gets inserted into the text. If I continue to hold `Caps-lock` while pressing `b` repeatedly, I get the expected behavior: the cursor moves backwards by words, leaving me in insert mode. Note that if I'm slow and press (and hold) `Caps-lock` for a while before pressing `b`, I see "<C->" appear on the screen when `Caps-lock` is initially pressed, but when I hit `b` iVim inserts *multiple* "<C->"s and *then* jumps back one word (so that the cursor is between the final "C" and "-"). It thus looks like a key repeat is happening in the background when `Caps-lock` is pressed and held.

4. Notice that my final comment on (3) about key repeat with `Caps-lock` might explain why multistage input with `ctrl` is not working for me (as noted in (1)(iii)).

5. I take it you meant to say that this solution has *not* been included in this build. (At least I'm not seeing it.)

**Comment by mvolkmann** [comment]
The idoc command paired with "Working Copy" works great!
Should the ":e" command work to reload iVim with the latest version of a file in the local Git repo? It doesn't seem to work.

---
R. Mark Volkmann
Object Computing, Inc.

**Comment by terrychou** [comment]
Submitted build 3 of the beta 1.12.

1. Except for chords `<A-Esc>`, `<A-arrows>`, `<A-Tab>` and `<A-CR>`, now my code does not handle the `alt` key. It means it should behave in the default way. However, there is still a problem I have found: the `<A-e>a` would leave two characters, therefore, it needs special handling.

2. The `<C-t>m` related problems have been solved.

3. By "`<C-CR>` never worked on my side", I mean the iOS system never gives me its event even though I have registered for it. It has nothing to do with the vim side, because it never reached my handling code. The chords `caps lock-key` have this same problem too. However, I have been testing it in the simulator. Please check whether they are caught on your devices or not.

4. I think we have a disagreement on the meaning of "multistage input". By it, I mean the input methods for languages such as Chinese or Japanese, whose characters normally need more than one key stroke to input. If I register the events of the letter keys stroke and handle them in my code (the key repeating would work if I do), the input of those languages will stop working. Therefore, repeating needs to give way.

5. As to the `caps lock` mapping feature, I don't have a good solution for the problems I mentioned before. Moreover, as said in 3, the iOS system even doesn't provide their events to my code. So, it will be removed for now.

6. The `delete` key now is interpretted as `<BS>` rather than `<C-h>`. I think this would solve the problem. (issue #17)

**Comment by bwhelm** [comment]
1. I'm seeing the same thing with `A-[normal key]`: typing `<A-e>a` produces "´é".  However, `A-Esc`, `A-arrows`, `A-Tab`, and `A-CR` all seem to work (as they do for you).

2. `<C-t>m` does work now -- thanks!

3. I'm not sure whether I'm seeing the same as you with `caps-lock` when this has been mapped to `ctrl`. With this mapping, pressing (and holding) `caps-lock` while typing another key no longer inserts `<C->` into the text, and iVim accepts the relevant control code. (Note: I've edited my previous comment under (3) to make visible text GitHub was interpreting as HTML code. Please look at it again.) So this is an improvement. However, when this happens, the `caps-lock` function is also triggered, which is not wanted. Note, however, that with `caps-lock` mapped to `ctrl`, I *can* enter `C-CR`, which I cannot do with the normal `ctrl` key.

4. Yes: I misunderstood you. I wonder if it might be useful to look at <https://github.com/blinksh/blink>, which does successfully make key repeating work. (Perhaps that's because they don't allow multistage input, but if they do have a genuine solution....)

5. Again, you might look at <https://github.com/blinksh/blink>, which has this working correctly.

6. Yes: this is working properly now.

Thanks again for your work on this: it is much improved!

**Comment by terrychou** [comment]
Tried the blink shell: (also in the simulator)

1. `<C-CR>` seems not working either.

2. Mapping `caps lock` to `esc` only works when `Auto repeat` is enabled. Otherwise, it will have the "uppercased letters" problem too. (by the way, setting in "MODIFIER MAPPINGS" didn't make the mapping work; "Caps as Esc" in "SPECIAL KEYS" did)

3. Setting to map `caps lock` to `ctrl` in "MODIFIER MAPPINGS" didn't work. Or is this the right way to do the mapping? Can you tell me the steps to make `caps lock -> ctrl` mapping work in blink?

4. No, blink shell doesn't support multistage input. The way in which the key repeating and `caps lock` mapping are implemented is at the cost of its absence.

**Comment by bwhelm** [comment]
Alas. Thanks for looking.

For (3): Settings > Keyboard > Modifier Mappings > CapsLock, select Ctrl. (That sounds like what you did, but it works for me on the iPad.

**Comment by terrychou** [comment]
@mvolkmann Could you open another issue for this problem? and please tell me the steps how you can reproduce it.

**Comment by holzschu** [comment]
It's great that the Escape and Control keys on external keyboards now work. 
I noticed that the function keys (F1, F2...) don't work. Can you enable them?

**Comment by holzschu** [comment]
Also, auto-repeat is really useful in command mode. I understand your issue with entering multiple-letter characters (although an option to disable that would be nice), but in command mode, I like to press "k" several times to move up or down.

**Comment by mvolkmann** [comment]
In Settings > Keyboard on my iPad, I don't see "Modifier Mappings".

---
R. Mark Volkmann
Object Computing, Inc.

**Comment by terrychou** [comment]
@holzschu Couldn't figure out how to register for events from the function keys. So they are not working (:h ios-external-keyboard).

As to the key repeating, I would wait for a better solution because the current candidate would bring confusion  to the whole app. (Incidently, IMHO, vim is the last place that needs this feature.)

**Comment by terrychou** [comment]
@bwhelm Tried it on the device too. It sort of "works". But since I have enabled several keyboards (English, Chinese, emoji...) and the original role of `caps lock` key is to switching among them, it will behave like: every time I trigger a `C-` chords through it, it will change my keyboard, then I need to switch it back manually. Sigh.

**Comment by bwhelm** [comment]
You found that remapping `CapsLock` to either `Ctrl` or `Esc` is incompatible with multistage input, and so made the quite reasonable decision to drop `CapsLock` remapping.

The thought occurred to me: might it be possible (and worthwhile) to create an option in Settings.app for a user to decide whether to enable `CapsLock` setting (at the expense of multistage input and keyboard switching) or disable it (to allow multistage input and keyboard switching)?

**Comment by terrychou** [comment]
I will think about it but with a low priority.

**Comment by mchalkley** [comment]
Rather than open a new issue, since this relates to an external keyboard, I thought I’d post it here, but my problem is VERY basic:  If I type :help ios immediately after opening the app, then ‘j’ down to see the lines further down in the file, each press of the ‘j’ key after reach the bottom of the buffer window increases the font size of the type, rather than scrolling down (except that it does show the one single next line in the file on the bottom line of the buffer window.  I’m using the latest version of the app on an iPad Pro 10.5...

Thanks,

Mark

**Comment by bwhelm** [comment]
I cannot duplicate this on an iPad Air2, with everything updated. (I tried both with my normal .vimrc/plugins and with `-u NONE` when launching iVim.)

**Comment by mchalkley** [comment]
It's pretty strange, but it happens with two different bluetooth keyboards on my Pro 10.5.  My Air2 doesn't have this behavior with either of the same two keyboards.  Neither iPad has any of iVim's default configurations modified (indeed, I'm not sure how to do that yet - I've been using the Buffers app for my Vim mode editing).

**Comment by mchalkley** [comment]
Never mind - it seems to have been caused by the Logitech keyboard driver.  I unpaired that keyboard and the problem disappeared.  I guess it's something specific to the 10.5...  I'm truly sorry for wasting your time.

**Comment by steakknife** [comment]
With a BT Magic Keyboard paired to an iOS device.. all the keys, except function keys (which may or may not be possible to grab events of), work OOTB. :+1:

**Comment by clarkezone** [comment]
wondering.. is it possible to implement autorepeat when on an external keyboard?  Particularly for cursor keys

**Comment by steakknife** [comment]
@clarkezone Yes, I would like to see that also. Having no keyboard repeat is too much like using a dial-up modem and makes iVim too impractical.
